includes/class-amp-service-worker.php

```diff
@@ -243,27 +243,20 @@ public static function get_precached_script_cdn_urls() {
 	 * Add hooks to install the service worker from AMP page.
 	 */
 	public static function add_install_hooks() {
-		if ( ! amp_is_legacy() && amp_is_request() ) {
-			add_action( 'wp_footer', [ __CLASS__, 'install_service_worker' ] );
-
-			// Prevent validation error due to the script that installs the service worker on non-AMP pages.
-			foreach ( [ 'wp_print_scripts', 'wp_print_footer_scripts' ] as $action ) {
-				$priority = has_action( $action, 'wp_print_service_workers' );
-				if ( false !== $priority ) {
-					remove_action( $action, 'wp_print_service_workers', $priority );
-				}
+		if ( ! amp_is_request() ) {
+			return;
+		}
+
+		// Prevent validation error due to the script that installs the service worker on non-AMP pages.
+		foreach ( [ 'wp_print_scripts', 'wp_print_footer_scripts' ] as $action ) {
+			$priority = has_action( $action, 'wp_print_service_workers' );
+			if ( false !== $priority ) {
+				remove_action( $action, 'wp_print_service_workers', $priority );
 			}
 		}
 
-		// Reader mode integration.
+		add_action( 'wp_footer', [ __CLASS__, 'install_service_worker' ] );
 		add_action( 'amp_post_template_footer', [ __CLASS__, 'install_service_worker' ] );
```

Comment:
Is it guaranteed that these two actions can never happen together for the same request? I'm thinking about something hacking the legacy template to add `wp_footer` to them as well...

I'd assume that that would cause havoc with the Reader mode, so not sure that is even a thing...

Reply:
These should be mutually-exclusive. If we want we could only do `amp_post_template_footer` if `amp_is_legacy()` but it seems overkill when the action will only ever be invoked in legacy post templates.